mirrord/intproxy/src/proxies/simple.rs

```diff
@@ -187,3 +206,43 @@ impl BackgroundTask for SimpleProxy {
         Ok(())
     }
 }
+
+/// Cache for successful [`DnsLookup`]s done by the agent.
+/// DNS mappings should not change very often, so it should be safe to cache these for the time of
+/// one mirrord session.
+#[derive(Default)]
+struct AddrInfoCache {
+    /// [`GetAddrInfoRequest::node`]s from the DNS requests that still require a response from the
+    /// agent.
+    outstanding_requests: VecDeque<String>,
+    /// Agent's successful responses to previous requests.
+    responses: HashMap<String, DnsLookup>,
+}
+
+impl AddrInfoCache {
+    /// Notifies this cache that a new [`GetAddrInfoRequest`] has been sent.
+    fn request_sent(&mut self, host: String) {
+        self.outstanding_requests.push_back(host);
+    }
+
+    /// Notifies this cache that the agent has responded with success to a [`GetAddrInfoRequest`].
+    /// The given [`DnsLookup`] is matched against the `host` from the oldest incomplete
+    /// [`GetAddrInfoRequest`] this struct knows of.
+    ///
+    /// # Panic
+    ///
+    /// This method panics if this struct does not now of any incomplete [`GetAddrInfoRequest`].
+    fn response_came(&mut self, lookup: DnsLookup) {
+        let host = self
+            .outstanding_requests
+            .pop_front()
+            .expect("received too many GetAddrInfoResponses");
```

Comment:
Does this ever happen?

Wouldn't it be better if this was just a noop in this case? (maybe also adding a log so we could debug it)

Something like:

```rust
fn foo() -> Option<()> {
   let host = foo.pop_front()?;
    ...

    Some(()) // and we never check this anywhere
}
```

Reply:
This can literally never happen :x

Reply:
Not with that attitude!